### PR TITLE
Return error if Transaction contains writable executable or ProgramData accounts

### PR DESCRIPTION
**Pull request by CriesofCarrots**
#### Problem
Accounts marked executable can only be written to if being upgraded (ie. if the UpgradeableLoader is present in the transaction). As of #19593, we're demoting top-level program ids incorrectly marked as writable, but we can't detect other executables until the accounts are loaded.
Similarly, ProgramData accounts can only be written if the UpgradeableLoader is present in the transaction, but we can't identify these accounts until they are loaded.

#### Summary of Changes
Return `InvalidAccountIndex` error to fail these types of Transactions and release the write lock quickly. Open to creating a new error type, if this seems significant enough.

Needs rebase on #19593 
@jackcmay  @jstarry , the last 3 commits are ready for review

### CODE REVIEW

**Comment by jackcmay** [comment]
Last 3 commits look good so far!

File: runtime/src/accounts.rs [hunk 1]
```diff
@@ -278,6 +278,14 @@ impl Accounts {
                             })
                             .unwrap_or_default();
 
+                        if account.executable()
+                            && demote_program_write_locks
+                            && message.is_writable(i, demote_program_write_locks)
+                        {
+                            error_counters.invalid_account_index += 1;
+                            return Err(TransactionError::InvalidAccountIndex);
+                        }
```
Comment (jstarry) [inline runtime/src/accounts.rs]: Does this mean we can't write lock upgradeable programs during upgrades? I think we need to add a check to see if the upgradeable loader is present

File: runtime/src/accounts.rs [hunk 2]
```diff
@@ -252,6 +252,7 @@ impl Accounts {
 
                     if solana_sdk::sysvar::instructions::check_id(key) {
                         if message.is_writable(i, demote_program_write_locks) {
+                            error_counters.invalid_account_index += 1;
```
Comment (jstarry) [inline runtime/src/accounts.rs:255]: the message writable check will always return false for the instructions sysvar so this will never be hit anyways fyi

File: runtime/src/accounts.rs (hunk 1 above)
Comment (CriesofCarrots) [reply runtime/src/accounts.rs]: Oh duh, I even said that myself in the slack thread. Thanks for catching me!

Comment (CriesofCarrots) [reply runtime/src/accounts.rs]: Fixed in updated https://github.com/solana-labs/solana/pull/19629/commits/08eed57dd468c7f9d8ea6058e09fe58729652234

File: runtime/src/accounts.rs (hunk 2 above)
Comment (CriesofCarrots) [reply runtime/src/accounts.rs:255]: True, I guess we should remove that check entirely?

Comment (jstarry) [reply runtime/src/accounts.rs:255]: Yeah, I'm in favour

Comment (CriesofCarrots) [reply runtime/src/accounts.rs]: ✅

**Review by KeKo6988 — APPROVED** [review]

**Comment by CriesofCarrots** [comment]
Ready for final review!

File: runtime/src/accounts.rs [hunk 3]
```diff
@@ -278,26 +275,49 @@ impl Accounts {
                             })
                             .unwrap_or_default();
 
-                        if account.executable() && bpf_loader_upgradeable::check_id(account.owner())
+                        if account.executable()
+                            && demote_program_write_locks
+                            && message.is_writable(i, demote_program_write_locks)
+                            && !is_upgradeable_loader_present(message)
```
Comment (sakridge) [inline runtime/src/accounts.rs:281]: can we hoist `is_upgradeable_loader_present` outside this loop?

Comment (CriesofCarrots) [reply runtime/src/accounts.rs:281]: Yes, ofc, thanks!

File: runtime/src/accounts.rs [hunk 4]
```diff
@@ -278,26 +275,49 @@ impl Accounts {
...
-                            if let Ok(UpgradeableLoaderState::Program {
-                                programdata_address,
-                            }) = account.state()
-                            {
-                                if let Some(account) = self
-                                    .accounts_db
-                                    .load_with_fixed_root(ancestors, &programdata_address)
-                                    .map(|(account, _)| account)
+                            error_counters.invalid_account_index += 1;
+                            return Err(TransactionError::InvalidAccountIndex);
```
Comment (jackcmay) [inline runtime/src/accounts.rs:284]: Would be helpful to return a new error here that better represents the failure

File: runtime/src/accounts.rs [hunk 5]
```diff
@@ -278,26 +275,49 @@ impl Accounts {
...
+                                    }
                                 } else {
-                                    error_counters.account_not_found += 1;
-                                    return Err(TransactionError::ProgramAccountNotFound);
+                                    error_counters.invalid_program_for_execution += 1;
+                                    return Err(TransactionError::InvalidProgramForExecution);
+                                }
+                            }
+
+                            // If a ProgramData account is present, it should only be writable
```
Comment (jackcmay) [inline runtime/src/accounts.rs:310]: Could this move into the success case of `load_with_fixed_root` above?

File: runtime/src/accounts.rs (hunk 4 above)
Comment (CriesofCarrots) [reply runtime/src/accounts.rs:284]: Yeah, I went back and forth on that, and ultimately went for less churn. But I'll add a unique error and error_counter

File: runtime/src/accounts.rs (hunk 5 above)
Comment (jackcmay) [reply runtime/src/accounts.rs:310]: Hmm, this is the 2nd case so no, but it could probably be inside an else so the account is not deserialized twice?

Comment (CriesofCarrots) [reply runtime/src/accounts.rs:310]: Oh good, I see your 2nd note now. I was stuck trying to figure out how it would work in `load_with_fixed_root`. Yes, great point, I'll rewrite to deser once.

File: runtime/src/accounts.rs [hunk 6]
```diff
@@ -278,26 +276,50 @@ impl Accounts {
...
-                                    return Err(TransactionError::ProgramAccountNotFound);
+                                    error_counters.invalid_program_for_execution += 1;
+                                    return Err(TransactionError::InvalidProgramForExecution);
                                 }
                             } else {
-                                error_counters.invalid_program_for_execution += 1;
-                                return Err(TransactionError::InvalidProgramForExecution);
+                                // If a ProgramData account is present, it should only be writable
+                                // if the bpf_loader_upgradeable account is also present
+                                if let Ok(UpgradeableLoaderState::ProgramData { .. }) =
```
Comment (jackcmay) [inline runtime/src/accounts.rs:312]: Would it be more efficient to first check the books and then deserialize?

File: runtime/src/accounts.rs [hunk 7]
```diff
@@ -251,9 +252,6 @@ impl Accounts {
                     }
 
                     if solana_sdk::sysvar::instructions::check_id(key) {
-                        if message.is_writable(i, demote_program_write_locks) {
```
Comment (jackcmay) [inline runtime/src/accounts.rs:254]: Why was this removed?

Comment (CriesofCarrots) [reply runtime/src/accounts.rs:254]: @jstarry pointed out that, because sysvar write locks are being demoted, this will never return true `if solana_sdk::sysvar::instructions::check_id(key)`.

Comment (CriesofCarrots) [reply runtime/src/accounts.rs:254]: So just tidying up.

File: runtime/src/accounts.rs [hunk 8]
```diff
@@ -278,26 +276,48 @@ impl Accounts {
                             })
                             .unwrap_or_default();
 
-                        if account.executable() && bpf_loader_upgradeable::check_id(account.owner())
```
Comment (jstarry) [inline runtime/src/accounts.rs:281]: Looks like the owner check was removed in order to apply this write restriction to all executable accounts. I think it would be better to distinguish between non-upgradeable programs and upgradeable programs though to make it impossible to write lock the token program.

File: runtime/src/accounts.rs (hunk 6 above)
Comment (jstarry) [reply runtime/src/accounts.rs:315]: Could we skip the deserialization altogether and disallow writes to any account owned by the upgradeable loader? The only time a program other than the upgradeable loader can write to one of these accounts is if they top up the lamports which doesn't seem useful anyways.

File: runtime/src/accounts.rs (hunk 8 above)
Comment (CriesofCarrots) [reply runtime/src/accounts.rs:281]: Oh, only include the `is_upgradeable_loader_present` exception if the owner is the upgradeable loader? Yeah, makes sense

**Comment by jackcmay** [comment]
Adding more lamports might be necessary when we add support for realloc...

**Comment by jstarry** [comment]
True, do you think the bpf loader program could be responsible for that?

**Comment by jackcmay** [comment]
I think so, the current plan is continue funding the account via the buffer account, so realloc to larger ProgramData account would require a larger buffer account with a larger associated number of tokens

**Comment by CriesofCarrots** [comment]
I'm not clear, @jackcmay , are you saying the upgradeable-loader program need to be present to do that sort of realloc?
If so, then I think we can skip the deserialization, because any type of `UpgradeableLoaderState` woudl require the program be present for writes.

**Comment by jackcmay** [comment]
In order for a program to be realloc'd larger the ProgramData account's balance will have to increase in order to still be rent-exempt.  The question was who/how transfers those lamports.  The current plan is for the larger buffer account to ferry along those lamports which means the upgradeable loader will always exist for all program reallocs.  aka, we can probably skip deserialization l like Justin suggested.

**Comment by CriesofCarrots** [comment]
Okay, the last commit changes the logic to return an error if:
- A writable account is owned by the upgradeable loader, but the upgradeable loader is not included in the tx
- OR, a writable account is executable

Look sufficient?
